src/core/models.py

```diff
@@ -1919,3 +1968,641 @@ def log_hijack_ended(sender, hijacker, hijacked, request, **kwargs):
+        submission_models.FrozenAuthor,
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
+    )
+    preprint_author = models.ForeignKey(
+        repository_models.PreprintAuthor,
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
+    )
+    title = models.CharField(
+        blank=True,
+        max_length=300,
+        verbose_name=_('Title, position, or role'),
+    )
+    department = models.CharField(
+        blank=True,
+        max_length=300,
+        verbose_name=_('Department, unit, or team'),
+    )
+    organization = models.ForeignKey(
+        'Organization',
+        blank=True,
+        null=True,
+        on_delete=models.CASCADE,
+    )
+    is_primary = models.BooleanField(
+        default=False,
+        help_text="Each account can have one primary affiliation",
+    )
+    start = models.DateField(
+        blank=True,
+        null=True,
+        verbose_name="Start date",
+    )
+    end = models.DateField(
+        blank=True,
+        null=True,
+        verbose_name="End date",
+        help_text="Leave empty for a current affiliation",
+    )
+
+    class Meta:
+        constraints = [
+            check_exclusive_fields_constraint(
+                ['account', 'frozen_author', 'preprint_author']
+            )
+        ]
+        ordering = ['is_primary', '-pk']
+
+    def title_department(self):
+        elements = [
+            self.title,
+            self.department,
+        ]
+        return ', '.join([element for element in elements if element])
+
+    def __str__(self):
+        elements = [
+            self.title,
+            self.department,
+            str(self.organization) if self.organization else '',
+        ]
+        return ', '.join([element for element in elements if element])
+
+
+    @property
+    def is_current(self):
+        if self.start and self.start > timezone.now():
+            return False
+        if self.end and self.end < timezone.now():
+            return False
+        return True
+
+    @classmethod
+    def set_primary_if_first(cls, obj):
+        other_affiliations = cls.objects.filter(
+            account=obj.account,
+            frozen_author=obj.frozen_author,
+            preprint_author=obj.preprint_author,
+        ).exclude(
+            pk=obj.pk
+        ).exists()
+        if not other_affiliations:
+            obj.is_primary = True
+
+    @classmethod
+    def keep_is_primary_unique(cls, obj):
+        if obj.is_primary:
+            cls.objects.filter(
+                is_primary=True,
+                account=obj.account,
+                frozen_author=obj.frozen_author,
+                preprint_author=obj.preprint_author,
+            ).exclude(
+                pk=obj.pk
+            ).update(
+                is_primary=False,
+            )
+
+    @classmethod
+    def get_primary(
+        cls,
+        account=None,
+        frozen_author=None,
+        preprint_author=None,
```

Comment:
I don't think we ever want three arguments and instead we can use a single mixed-typed argument (e.g. 'author') which can be an instance of a Model that has the `affiliation_set` interface